common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala

```diff
@@ -57,3 +65,23 @@ trait ContainerFactoryProvider extends Spi {
                           instance: InstanceId,
                           parameters: Map[String, Set[String]]): ContainerFactory
 }
+
+object ContainerFactory {
+
+  val containerArgsConfig = {
+    //since typesafe (lightbend) Config does NOT support parsing complex objects from system properties, we need a ConfigReader
+    //see https://github.com/lightbend/config/blob/master/HOCON.md#java-properties-mapping for details on properties handling from typesafe Config
+    //see pureconfig docs at https://pureconfig.github.io/docs/combinators.html
```

Comment:
Can you elaborate on this a bit?

tried this locally:

```scala
case class ContainerArgsConfig(network: String, dnsServers: Seq[String], extraArgs: Map[String, Set[String]])

object Testing {
  import pureconfig._

  def main(args: Array[String]): Unit = {
    val config = """
                   |network: bridge
                   |dns-servers: [google.com]
                   |extra-args: {
                   |  test: [a, b]
                   |  another: [c, d]
                   |}
                   | """.stripMargin

    println(loadConfigOrThrow[ContainerArgsConfig](ConfigFactory.parseString(config)))
  }
}
```

Which yields:

```
ContainerArgsConfig(bridge,List(google.com),Map(another -> Set(c, d), test -> Set(a, b)))
```

Do we need something different?

Reply:
I tried to be as specific as possible in the comment :)

It is this: _typesafe (lightbend) Config does NOT support parsing complex objects from system properties_ i.e. when you override values via system properties (e.g. the `CONFIG_*` env vars in our case), they are always read as String type, never converted to List/Map/Object/etc.

So, in your example, you did not override the config from system properties.

IMHO this is a real flaw in lightbend/typesafe config, and we just need to be very careful about what types are parsed from the config, and make sure the override via system property works as you expect. I saw some similar logged issues with lightbend config, along with some arguments as to why it won't be fixed...

In fact, I just notice the same flaw exists for the `dns-servers` property - I'll add a config reader for that as well. Now I'm also wonder if the DockerClientTimeoutConfig will exhibit the same issue, if you try to override via system properties.

Reply:
Actually for 'dns-servers' (type `Seq[String]`), there exists a way to handle this via typesafe java properties parser: https://github.com/lightbend/config/pull/70

You have to specify each element in the list as a separate system property whose key ends with the index number in the list, e.g.
```
    System.setProperty("whisk.container-factory.container-args.dns-servers.0", "google.com")
    System.setProperty("whisk.container-factory.container-args.dns-servers.1", "1.2.3.4")
```

Note that we _could_ alternatively choose to add a custom parser for this, so that we can use a single property like "dns-servers"="[google.com,1.2.3.4]", but I ran into this issue trying to use the typesafe parser https://github.com/lightbend/config/issues/460 so we'd need to parse the value in yet another way.

Reply:
Got it, that makes sense!

Actually using your workaround for the lists, overriding the extra args works exactly the same way:

```scala
case class ContainerArgsConfig(network: String, dnsServers: Seq[String], extraArgs: Map[String, Set[String]])

object Testing {
  import pureconfig._

  def main(args: Array[String]): Unit = {
    System.setProperty("network", "test")
    System.setProperty("dns-servers.0", "google.com")
    System.setProperty("extra-args.test.0", "a")
    System.setProperty("extra-args.test.1", "b")
    System.setProperty("extra-args.try.0", "a")

    println(loadConfigOrThrow[ContainerArgsConfig](""))
  }
}
```